x/rewards/keeper/distribution.go

```diff
@@ -0,0 +1,231 @@
+// Func iterates over all tracked transactions and estimates gas usage for each contract (on block and tx levels) merging operations.
+func (k Keeper) estimateBlockGasUsage(ctx sdk.Context, height int64) *blockRewardsDistributionState {
+	metadataState := k.state.ContractMetadataState(ctx)
+
+	// Get all tracked transactions by the x/tracking module
+	blockGasTrackingInfo := k.trackingKeeper.GetBlockTrackingInfo(ctx, height)
+
+	// Create a new block rewards distribution state and fill it up
+	blockDistrState := &blockRewardsDistributionState{
+		Height:    height,
+		Txs:       make(map[uint64]uint64, len(blockGasTrackingInfo.Txs)),
+		Contracts: make(map[string]*contractRewardsDistributionState, 0),
+	}
+
+	// Fill up gas usage iterating over all tracked transactions and contract operations
+	for _, txGasTrackingInfo := range blockGasTrackingInfo.Txs {
+		// Set total gas used by the transaction
+		blockDistrState.Txs[txGasTrackingInfo.Info.Id] = txGasTrackingInfo.Info.TotalGas
+
+		// Estimate contract operations total gas used for this transaction
+		for _, contractOp := range txGasTrackingInfo.ContractOperations {
+			opGasUsed, opEligible := contractOp.GasUsed()
+			if !opEligible {
+				// Skip noop operation (should not happen since we're tracking an actual WASM usage)
+				k.Logger(ctx).Debug("Noop contract operation found (skip)", "txID", contractOp.TxId, "opID", contractOp.Id)
+				continue
+			}
+
+			// Create a new contract rewards distribution state
+			contractDistrState := blockDistrState.Contracts[contractOp.ContractAddress]
+			if contractDistrState == nil {
+				contractDistrState = &contractRewardsDistributionState{
+					ContractAddress:     contractOp.MustGetContractAddress(),
+					TxGasUsed:           make(map[uint64]uint64, 0),
+					InflationaryRewards: sdk.Coin{Amount: sdk.ZeroInt()}, // necessary to avoid nil pointer panic on Coins.Add call
+				}
+				if metadata, found := metadataState.GetContractMetadata(contractDistrState.ContractAddress); found {
+					contractDistrState.Metadata = &metadata
+				}
+				blockDistrState.Contracts[contractOp.ContractAddress] = contractDistrState
+			}
+
+			// Increase block gas usage
+			contractDistrState.BlockGasUsed += opGasUsed
+
+			// Increase tx gas usage
+			txGasUsed := contractDistrState.TxGasUsed[contractOp.TxId] // 0 if not initialized
+			contractDistrState.TxGasUsed[contractOp.TxId] = txGasUsed + opGasUsed
+		}
+	}
+
+	return blockDistrState
+}
+
+// estimateBlockRewards update block distribution state with tracked rewards calculating reward shares per contract.
+// Func iterates over all tracked transactions and estimates inflation (on block level) and fee rebate (merging
+// tokens for each transaction contract has operation at) rewards for each contract.
+func (k Keeper) estimateBlockRewards(ctx sdk.Context, blockDistrState *blockRewardsDistributionState) *blockRewardsDistributionState {
```

Comment:
You are right about pointers in general. Goway is to be as immutable as possible. But here, since we are not "escaping" (not in terms of the compiler) the `DistributeRewards` func, this is kind of OK. But if that really bothers you, I can pass an object and return an updated one. BTW maps in the State object are "heaped" (probably 🤷‍♂️).

Reply:
Let's not get ahead of ourselves here, I'll just keep a mental note on it as we observe further growth.

P.D: yeah maps are notoriously bad in golang for memory growth more on this [golang issue](https://github.com/golang/go/issues/20135)